### PR TITLE
add scratch var scope with reuse==False

### PR DESCRIPTION
**Pull request by berleant**
Just started with tensorflow yesterday, so my confidence that this is the right/best fix is low. However, the `build_model()` step would not complete pre-commit, and now it does.

I'm opening this PR so I can get feedback. Thanks!

https://github.com/bamos/dcgan-completion.tensorflow/issues/14

### CODE REVIEW

**Comment by berleant** [comment]
It isn't compatible with the checkpoint file that comes with the repository, but otherwise it's working well for me.

**Comment by mgarbade** [comment]
This fix works for me too (TF 0.12.1). Could you explain what it does?

**Comment by berleant** [comment]
Sure! I recommend reading the official documentation on varscopes and the "reuse" keyword. That's where it all came together for me.

There are two discriminator models (one for true samples, one for generated samples), called `D` and `D_`. `D_` is declared with the `reuse` parameter (https://github.com/bamos/dcgan-completion.tensorflow/blob/bee69486df75ba324acc54b86cc48dcee26e0698/model.py#L81). The `reuse` parameter is inherited, so it applies to everything in the `D_` scope, including `batch_norm`. When `reuse` is `True`, TensorFlow does not allow you to create new variables. In `batch_norm`, `self.ema.apply()` always creates a new variable within itself, which trips the error, because `reuse` is `True`. So I pre-declared a scratch varscope with `reuse` being `False` (`SCRATCH_VARSCOPE`), and used the handy `with` statement to break temporarily into that varscope before calling `apply()`, which is then free to create as many variables as it wants.

I am 90% sure this is a correct fix, because `self.ema.apply()`'s new variables are used immediately and then never again. In the worst case, no batch normalization occurs, but at least it runs!

Potential issue: I *think* that the scratch varscope will grow indefinitely, with new variables declared every time we perform batch normalization. Could become an issue eventually, in theory, but for a tutorial it's probably fine.

(It used to work because the behavior of `reuse` was different/incorrect. Someone did make an issue requesting that `self.ema.apply()` (`tf.train.ExponentialMovingAverage()`) not create new variables, because they ran into the same problem, but the issue was closed. You should be able to find the issue if you google.)

**Comment by anshitm** [comment]
It solved the problem for me too but now the checkpoint file in the repository dosen't work . Can you please update them in your repository. I am having an amd gpu, training will take a infinite time over cpu.

**Comment by bamos** [comment]
Hi, I've merged in #21 instead, which adds support for TF 1.0
